ldelf/ta_elf.c

```diff
@@ -849,64 +847,128 @@ static void close_handle(struct ta_elf *elf)
 	elf->handle = -1;
 }
 
-void ta_elf_load_main(const TEE_UUID *uuid, uint32_t *is_32bit,
-		      uint64_t *entry, uint64_t *sp, uint32_t *ta_flags)
+static void clean_elf_load_main(struct ta_elf *elf)
 {
-	struct ta_elf *elf = queue_elf(uuid);
-	struct ta_head *head;
-	vaddr_t va = 0;
 	TEE_Result res = TEE_SUCCESS;
 
-	assert(elf);
-	elf->is_main = true;
-
-	init_elf(elf);
-
 	/*
-	 * Legacy TAs doesn't set entry point, instead it's set in ta_head.
-	 * If entry point isn't set explicitly, set to the start of the
-	 * first executable section by the linker. Since ta_head also
-	 * always comes first in legacy TA it means that the entry point
-	 * will be set to 0x20.
-	 *
-	 * NB, everything before the commit a73b5878c89d ("Replace
-	 * ta_head.entry with elf entry") is considered legacy TAs for
-	 * ldelf.
+	 * Clean up from last attempt to load
 	 */
-	if (elf->e_entry == sizeof(*head))
-		elf->is_legacy = true;
+	res = sys_unmap(elf->ehdr_addr, SMALL_PAGE_SIZE);
+	if (res)
+		err(res, "sys_unmap");
 
+	while (!TAILQ_EMPTY(&elf->segs)) {
+		struct segment *seg = TAILQ_FIRST(&elf->segs);
+		vaddr_t va = 0;
+		size_t num_bytes = 0;
+
+		va = rounddown(elf->load_addr + seg->vaddr);
+		if (seg->remapped_writeable)
+			num_bytes = roundup(seg->vaddr + seg->memsz) -
+				    rounddown(seg->vaddr);
+		else
+			num_bytes = seg->memsz;
+
+		res = sys_unmap(va, num_bytes);
+		if (res)
+			err(res, "sys_unmap");
+
+		TAILQ_REMOVE(&elf->segs, seg, link);
+		free(seg);
+	}
+
+	free(elf->shdr);
+	memset(&elf->is_32bit, 0,
+	       (vaddr_t)&elf->uuid - (vaddr_t)&elf->is_32bit);
+
+	TAILQ_INIT(&elf->segs);
+}
+
+static void load_main(struct ta_elf *elf)
+{
+	struct ta_head *head = NULL;
+
+	init_elf(elf);
 	map_segments(elf);
+	populate_segments(elf);
 	add_dependencies(elf);
 	copy_section_headers(elf);
 	save_symtab(elf);
 	close_handle(elf);
 
 	head = (struct ta_head *)elf->load_addr;
+	if (head->depr_entry != UINT64_MAX) {
+		/*
+		 * Legacy TAs sets their entry point in ta_head. For
+		 * non-legacy TAs the entry point of the ELF is set instead
+		 * and leaving the ta_head entry point set to UINT64_MAX to
+		 * indicate that it's not used.
+		 *
+		 * NB, everything before the commit a73b5878c89d ("Replace
+		 * ta_head.entry with elf entry") is considered legacy TAs
+		 * for ldelf.
+		 *
+		 * Legacy TAs cannot be mapped with shared memory segments
+		 * so restart the mapping if it turned out we're loading a
+		 * legacy TA.
+		 */
 
-	*is_32bit = elf->is_32bit;
-	if (elf->is_legacy) {
-		assert(head->depr_entry != UINT64_MAX);
-		*entry = head->depr_entry + elf->load_addr;
-	} else {
-		assert(head->depr_entry == UINT64_MAX);
-		*entry = elf->e_entry + elf->load_addr;
+		DMSG("Reloading TA %pUl as legacy TA", (void *)&elf->uuid);
```

Comment:
What if the TA changes on disk at this point, and is replaced by a non-legacy TA?

Reply:
Fixes